presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -192,31 +191,32 @@ public StageInfo getStageInfo()
                 ImmutableList::of);
     }
 
-    public synchronized void addExchangeLocation(ExchangeLocation exchangeLocation)
+    public synchronized void addExchangeLocations(PlanFragmentId fragmentId, Set<URI> exchangeLocations, boolean noMoreExchangeLocations)
     {
-        requireNonNull(exchangeLocation, "exchangeLocation is null");
-        RemoteSourceNode remoteSource = exchangeSources.get(exchangeLocation.getPlanFragmentId());
-        checkArgument(remoteSource != null, "Unknown remote source %s. Known sources are %s", exchangeLocation.getPlanFragmentId(), exchangeSources.keySet());
+        requireNonNull(noMoreExchangeLocations, "noMoreExchangeLocations is null");
 
-        exchangeLocations.put(remoteSource.getId(), exchangeLocation.getUri());
-        for (RemoteTask task : getAllTasks()) {
-            task.addSplits(remoteSource.getId(), ImmutableList.of(createRemoteSplitFor(task.getTaskInfo().getTaskId(), exchangeLocation.getUri())));
-        }
-    }
-
-    public synchronized void noMoreExchangeLocationsFor(PlanFragmentId fragmentId)
-    {
-        requireNonNull(fragmentId, "fragmentId is null");
         RemoteSourceNode remoteSource = exchangeSources.get(fragmentId);
         checkArgument(remoteSource != null, "Unknown remote source %s. Known sources are %s", fragmentId, exchangeSources.keySet());
 
-        completeSourceFragments.add(fragmentId);
+        this.exchangeLocations.putAll(remoteSource.getId(), exchangeLocations);
 
-        // is the source now complete?
-        if (completeSourceFragments.containsAll(remoteSource.getSourceFragmentIds())) {
-            completeSources.add(remoteSource.getId());
-            for (RemoteTask task : getAllTasks()) {
-                task.noMoreSplits(remoteSource.getId());
+        for (RemoteTask task : getAllTasks()) {
+            ImmutableMultimap.Builder<PlanNodeId, Split> newSplits = ImmutableMultimap.builder();
+            for (URI exchangeLocation : exchangeLocations) {
+                newSplits.put(remoteSource.getId(), createRemoteSplitFor(task.getTaskInfo().getTaskId(), exchangeLocation));
+            }
+            task.addSplits(newSplits.build());
```

Comment:
both `task.addSplits` and `task.noMoreSplits` calls scheduleUpdate. It is better now that it is async. But, since we have all the information here, should we create an `addSplits(Iterable<Split>splits, boolean noMoreSplits)` API in the `RemoteTask` and reduce an additional call to scheduleUpdate?